File: 0002-add-two-numbers/0002-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        r1 , r2 = [] ,[]
        c1 , c2 = l1 , l2 
        while c1 :
            r1.append(c1.val)
            c1 = c1.next
        while c2 :
            r2.append(c2.val)
            c2=c2.next
        r1.reverse()
        r2.reverse()
        r1 = ''.join(map(str , r1))
        r2 = ''.join(map(str ,r2)) 
        res = int(r1) +int(r2) 
        
        res = [int(digit) for digit in str(res)]
        res.reverse()
        head = ListNode(res[0])
        current = head 
        for val in res[1:]:
            current.next = ListNode(val)
            current = current.next

        return head

        

File: 0002-add-two-numbers/0002-add-two-numbers.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */


struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
    struct ListNode *l3 = (struct ListNode*)malloc(sizeof(struct ListNode));
    if (l3 == NULL) {
        printf("Memory allocation failed\n");
        exit(0);
    }

    struct ListNode *head = l3; // Pointer to track the current node
    int carry = 0;

    while (l1 != NULL || l2 != NULL || carry != 0) {
        int sum = carry; // Start with carry

        if (l1 != NULL) {
            sum += l1->val;
            l1 = l1->next;
        }
        
        if (l2 != NULL) {
            sum += l2->val;
            l2 = l2->next;
        }

        head->val = sum % 10; // Store the digit
        carry = sum / 10;     // Update the carry

        // Only create the next node if necessary
        if (l1 != NULL || l2 != NULL || carry != 0) {
            head->next = (struct ListNode*)malloc(sizeof(struct ListNode));
            if (head->next == NULL) {
                printf("Memory allocation failed\n");
                exit(0);
            }
            head = head->next;
            head->next = NULL; // Initialize the next pointer
        }
    }

    head->next = NULL; // Set the last node's next to NULL to avoid misalignment
    return l3;        // Return the head of the list
}
